go/vt/vttablet/tabletmanager/framework_test.go

```diff
@@ -0,0 +1,228 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tabletmanager
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"vitess.io/vitess/go/mysql/fakesqldb"
+	"vitess.io/vitess/go/sqltypes"
+	"vitess.io/vitess/go/vt/binlog/binlogplayer"
+	"vitess.io/vitess/go/vt/dbconfigs"
+	"vitess.io/vitess/go/vt/grpcclient"
+	"vitess.io/vitess/go/vt/mysqlctl"
+	"vitess.io/vitess/go/vt/topo"
+	"vitess.io/vitess/go/vt/topo/memorytopo"
+	"vitess.io/vitess/go/vt/vttablet/queryservice"
+	"vitess.io/vitess/go/vt/vttablet/tabletconn"
+	"vitess.io/vitess/go/vt/vttablet/tabletconntest"
+	"vitess.io/vitess/go/vt/vttablet/tabletmanager/vreplication"
+	"vitess.io/vitess/go/vt/vttablet/tmclient"
+	"vitess.io/vitess/go/vt/vttablet/tmclienttest"
+
+	querypb "vitess.io/vitess/go/vt/proto/query"
+	"vitess.io/vitess/go/vt/proto/tabletmanagerdata"
+	tabletmanagerdatapb "vitess.io/vitess/go/vt/proto/tabletmanagerdata"
+	"vitess.io/vitess/go/vt/proto/topodata"
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+)
+
+type testEnv struct {
+	mu         sync.Mutex
+	ctx        context.Context
+	vrengine   *vreplication.Engine
+	vrdbClient *binlogplayer.MockDBClient
+	ts         *topo.Server
+	cells      []string
+	tablets    map[int]*fakeTabletConn
+	tmc        *fakeTMClient
+	dbName     string
+	protoName  string
+}
+
+func newTestEnv(t *testing.T) *testEnv {
+	tenv := &testEnv{
+		ctx:        context.Background(),
+		vrdbClient: binlogplayer.NewMockDBClient(t),
+		tablets:    make(map[int]*fakeTabletConn),
+		tmc:        newFakeTMClient(),
+		cells:      []string{"zone1"},
+		dbName:     "tmtestdb",
+		protoName:  t.Name(),
+	}
+	tenv.mu.Lock()
+	defer tenv.mu.Unlock()
+	tenv.ts = memorytopo.NewServer(tenv.cells...)
+
+	tabletconn.RegisterDialer(t.Name(), func(tablet *topodatapb.Tablet, failFast grpcclient.FailFast) (queryservice.QueryService, error) {
+		tenv.mu.Lock()
+		defer tenv.mu.Unlock()
+		if qs, ok := tenv.tablets[int(tablet.Alias.Uid)]; ok {
+			return qs, nil
+		}
+		return nil, fmt.Errorf("tablet %d not found", tablet.Alias.Uid)
+	})
+	tabletconntest.SetProtocol(fmt.Sprintf("go.vt.vttablet.tabletmanager.framework_test_%s", t.Name()), tenv.protoName)
+	tmclient.RegisterTabletManagerClientFactory(t.Name(), func() tmclient.TabletManagerClient {
+		return tenv.tmc
+	})
+	tmclienttest.SetProtocol(fmt.Sprintf("go.vt.vttablet.tabletmanager.framework_test_%s", t.Name()), tenv.protoName)
+
+	dbClientFactory := func() binlogplayer.DBClient {
+		return tenv.vrdbClient
+	}
+	tenv.vrengine = vreplication.NewTestEngine(tenv.ts, tenv.cells[0], mysqlctl.NewFakeMysqlDaemon(fakesqldb.New(t)), dbClientFactory, dbClientFactory, tenv.dbName, nil)
+	tenv.vrdbClient.ExpectRequest(fmt.Sprintf("select * from _vt.vreplication where db_name='%s'", tenv.dbName), &sqltypes.Result{}, nil)
+	tenv.vrengine.Open(tenv.ctx)
+	require.True(t, tenv.vrengine.IsOpen())
```

Comment:
so if this were ever to fail, it would just say something like "assertion failed: false is not true"; we might want to add a message to be a bit more helpful to debugging